Support for CP String patching

fengxue-IS opened the pull request:
- New J9ClassPatchMap structure to store patching info
- Update defineClassCommon code to accept J9ClassPatchMap
- Patch RAM constantpool before returing class object

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

babsingh commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -1957,6 +1958,11 @@ typedef struct J9ClassPathEntry {
+typedef struct J9ClassPatchMap {
+	U_16 size;
+	U_16* indexMap;
+}
```
compilation error:
```suggestion
} J9ClassPatchMap;
```

babsingh commented:
@fengxue-IS i have verified that these changes work with JEP-370. can we remove the WIP and have the pull request merged?

fengxue-IS commented:
@babsingh can you take another look please?

babsingh commented on runtime/jcl/common/sun_misc_Unsafe.cpp:
```diff
@@ -96,19 +95,65 @@ Java_sun_misc_Unsafe_defineAnonymousClass(JNIEnv *env, jobject receiver, jclass
+
+				for (U_16 j = 1; j < clazz->romClass->ramConstantPoolCount; j++) {
+					if ((J9_CP_TYPE(cpShapeDescription, j) == J9CPTYPE_STRING)
+					&& (J9UTF8_EQUALS(romString, J9ROMSTRINGREF_UTF8DATA((J9ROMStringRef *)&romCP[j])))) {
```
```suggestion
					if ((J9_CP_TYPE(cpShapeDescription, j) == J9CPTYPE_STRING)
						&& J9UTF8_EQUALS(romString, J9ROMSTRINGREF_UTF8DATA((J9ROMStringRef *)&romCP[j]))
					) {
```

fengxue-IS replied:
fixed as per suggested

fengxue-IS commented:
I've compiled this change locally and passed JEP370 tests (assert null not fixed yet on my local branch).

@babsingh could you run the testing with the lastest JEP 370 branch please, thanks

marking this as ready for formal review

babsingh commented:
> @babsingh could you run the testing with the lastest JEP 370 branch please, thanks

These changes yield the correct behavior. No failures seen in JEP-370 testing. Refer to https://github.com/eclipse/openj9/issues/8292#issuecomment-582193068 for more details.

fjeremic commented:
The commit messages contain no details on what this change does. It would be helpful for readers such has myself who don't know much about what is being done here to put a description of the change.

fengxue-IS commented:
Added comments to the code/commit, added new null check after memory allocation.

This PR replaces #8369 , the old one will be kept as a WIP draft for future references.

@DanHeidinga can you please take a look
FYI @babsingh

DanHeidinga commented on runtime/bcutil/ConstantPoolMap.cpp:
```diff
@@ -302,6 +302,20 @@ ConstantPoolMap::computeConstantPoolMapAndSizes()
+	J9ClassPatchMap *map = _context->patchMap();
+
+	/**
+	 * If valid patchMap structure passed, class require ConstantPool patching,
```
```suggestion
	 * If a valid patchMap structure is passed, this class requirse ConstantPool patching.
```

DanHeidinga commented on runtime/bcutil/ConstantPoolMap.cpp:
```diff
@@ -302,6 +302,20 @@ ConstantPoolMap::computeConstantPoolMapAndSizes()
+
+	/**
+	 * If valid patchMap structure passed, class require ConstantPool patching,
+	 * record Classfile to J9Class constantpool index mapping to patch RAM Class
```
```suggestion
	 * Record the index mapping from Classfile to J9Class constantpool to allow patching the RAM CP
```

DanHeidinga commented on runtime/jcl/common/sun_misc_Unsafe.cpp:
```diff
@@ -96,19 +95,72 @@ Java_sun_misc_Unsafe_defineAnonymousClass(JNIEnv *env, jobject receiver, jclass
+			cpPatchMap.indexMap = (U_16 *)j9mem_allocate_memory(cpPatchMap.size * sizeof(U_16), J9MEM_CATEGORY_VM);
+
+			if (cpPatchMap.indexMap == NULL) {
+				vmFuncs->throwNativeOOMError(env, 0, 0);
```
This function gets VMAccess so it can't be called with it.  Need to swap these two lines

DanHeidinga commented on runtime/jcl/common/sun_misc_Unsafe.cpp:
```diff
@@ -96,19 +95,72 @@ Java_sun_misc_Unsafe_defineAnonymousClass(JNIEnv *env, jobject receiver, jclass
+		}
+
+		if (cpPatchMap.size > BUFFER_SIZE) {
+			j9mem_free_memory(cpPatchMap.indexMap);
```
It's a good practice to null out fields that have been released even when the end of the function is so close.  Future changes to the file may accidentally reuse the freed mem if we don't null it now.

DanHeidinga commented:
Jenkins test sanity zlinux jdk8,jdk14

DanHeidinga commented:
@fengxue-IS Can you record the build results / links before pushing new changes to this PR?  I'll be on a plane shortly and likely won't see the results in time.

pshipton commented:
The builds passed https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/2621/

fengxue-IS commented:
Updated code as suggested by review comments.

DanHeidinga commented:
Jenkins test sanity zlinux jdk14
